File: client/android/app/src/main/java/com/shelterfinder/app/FirebaseTopicsPlugin.java
package com.shelterfinder.app;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.firebase.messaging.FirebaseMessaging;

@CapacitorPlugin(name = "FirebaseTopics")
public class FirebaseTopicsPlugin extends Plugin {

    @PluginMethod
    public void subscribeToTopic(PluginCall call) {
        String topic = call.getString("topic");
        if (topic == null || topic.isEmpty()) {
            call.reject("Topic is required");
            return;
        }

        FirebaseMessaging.getInstance().subscribeToTopic(topic)
            .addOnSuccessListener(aVoid -> call.resolve())
            .addOnFailureListener(e -> call.reject("Subscribe failed: " + e.getMessage()));
    }

    @PluginMethod
    public void unsubscribeFromTopic(PluginCall call) {
        String topic = call.getString("topic");
        if (topic == null || topic.isEmpty()) {
            call.reject("Topic is required");
            return;
        }

        FirebaseMessaging.getInstance().unsubscribeFromTopic(topic)
            .addOnSuccessListener(aVoid -> call.resolve())
            .addOnFailureListener(e -> call.reject("Unsubscribe failed: " + e.getMessage()));
    }
}

File: client/ios/App/App/FirebaseTopicsPlugin.swift
import Foundation
import Capacitor
import FirebaseMessaging

/**
 * Bridges FCM topic subscription to JavaScript.
 * Mirrors the Android FirebaseTopicsPlugin so the shared TS code in
 * src/services/pushNotifications.ts works identically on both platforms.
 */
@objc(FirebaseTopicsPlugin)
public class FirebaseTopicsPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "FirebaseTopicsPlugin"
    public let jsName = "FirebaseTopics"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "subscribeToTopic", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unsubscribeFromTopic", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getToken", returnType: CAPPluginReturnPromise)
    ]

    @objc func subscribeToTopic(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic"), !topic.isEmpty else {
            call.reject("Topic is required")
            return
        }

        Messaging.messaging().subscribe(toTopic: topic) { error in
            if let error = error {
                call.reject("Subscribe failed: " + error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func unsubscribeFromTopic(_ call: CAPPluginCall) {
        guard let topic = call.getString("topic"), !topic.isEmpty else {
            call.reject("Topic is required")
            return
        }

        Messaging.messaging().unsubscribe(fromTopic: topic) { error in
            if let error = error {
                call.reject("Unsubscribe failed: " + error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    /// Returns the FCM registration token (not the raw APNs token).
    /// The server sends via FCM, so this is the value /api/register-token needs.
    @objc func getToken(_ call: CAPPluginCall) {
        Messaging.messaging().token { token, error in
            if let error = error {
                call.reject("Token fetch failed: " + error.localizedDescription)
            } else {
                call.resolve(["token": token ?? ""])
            }
        }
    }
}
